### PR TITLE
sstable: add meta block for range keys; support writing range keys

### PR DESCRIPTION
**Pull request by nicktrav**
Range keys (see #1341) will be stored in their own, single block of an
sstable. Add a new, optional meta block, indexed as "pebble.range_key"
in the metablock index, to the sstable structure. This block is only
present when at least one range key has been written to the sstable.

Add the ability to add range keys to an sstable via
`(*sstable.Writer).Write`.

Update existing data-driven tests to support printing of the range key
summary. Add additional test coverage demonstrating writing of range
keys with an `sstable.Writer`.

Add minimal functionality to `sstable.Reader` to support writing the
data-driven test cases for the writer. Additional read-oriented
functionality will be added in a subsequent patch.

Related to #1339.

### CODE REVIEW

File: testdata/event_listener [hunk 1]
```diff
@@ -183,7 +183,7 @@ compact         1   2.3 K     0 B       0          (size == estimated-debt, scor
 zmemtbl         0     0 B
    ztbl         0     0 B
  bcache         8   1.4 K    5.9%  (score == hit-rate)
- tcache         1   616 B    0.0%  (score == hit-rate)
+ tcache         1   640 B    0.0%  (score == hit-rate)
```
Comment (nicktrav) [inline testdata/event_listener:186]: These extra 24 bytes comes from the 16 bytes in the new range key block handle and the 8 bytes for the uint64 for the range key counter in the metadata.

File: sstable/writer.go [hunk 2]
```diff
@@ -370,6 +385,92 @@ func (w *Writer) addTombstone(key InternalKey, value []byte) error {
 	return nil
 }
 
+func (w *Writer) addRangeKey(key InternalKey, value []byte) error {
+	if !w.disableKeyOrderChecks && w.rangeKeyBlock.nEntries > 0 {
+		// Convert both the previous and current keys into spans so that we can
```
Comment (nicktrav) [inline sstable/writer.go:390]: Note to self, this comment is old. Update it.

File: sstable/writer.go [hunk 3]
```diff
@@ -370,6 +385,92 @@ func (w *Writer) addTombstone(key InternalKey, value []byte) error {
...
+			// key and value when it was first added to teh range key block.
+			panic(errors.Errorf("pebble: invalid end key for span: %s",
+				(keyspan.Span{Start: prevStartKey, End: prevEndKey}).Pretty(w.formatKey)))
+		}
+
+		curStartKey := key
+		curEndKey, curVal, ok := rangekey.DecodeEndKey(curStartKey.Kind(), value)
+		if !ok {
+			w.err = errors.Errorf("pebble: invalid end key for span: %s",
+				(keyspan.Span{Start: prevStartKey, End: prevEndKey}).Pretty(w.formatKey))
```
Comment (nicktrav) [inline sstable/writer.go:405]: Note to self, fix the span in the error message.

**Review by jbowens — COMMENTED** [review]
Took a quick pass: looks good! I think the big item remaining is the public interface, but we can also defer the interface to subsequent PRs.

> Reviewed 5 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400#-:-MqQYgq-5ahfJCTZ8TKf:b4gqoay)* status: 5 of 14 files reviewed, 8 unresolved discussions (waiting on @jbowens and @nicktrav)

___
*[sstable/table.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQFDUa3Il1B79v2gjT:-MqQFDUa3Il1B79v2gjU:bb7ohn6) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/table.go#L162)):*
> ```Go
> 	rocksDBFormatVersion2 = 2
> 
> 	// FIXME(travers): Are we ok with this naming?
> ```

sgtm
___
*[sstable/writer.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQS8GsCXH22tWjKSNd:-MqQS8GsCXH22tWjKSNe:b3rka88) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L245)):*
> ```Go
> 	}
> 	return w.addPoint(key, value)
> }
> ```

I alluded to this when we talked last week, but there's a complication to the interface that we export in that CockroachDB needs to be able to construct sstables for ingestion that contain range keys, but we don't want to expose the internals of fragmentation, coalescing or the value-encoding to CockroachDB.

When CockroachDB is constructing a sstable for ingest, it'll read individual fragmented range key sets (containing a single suffix, value pair and bounds). As it reads them, it'll need to pass them to the `Writer`. However, we don't want to just write those verbatim to the sstable block as internal range keys: we'll want to fragment and coalesce the logical range keys, so that we write coalesced range keys to the block.

On the other hand, during a compaction Pebble will use a merging range key iter to produce coalesced, nonoverlapping `CoalescedSpan`s that need to be written to the sstable as up to 1 key per range-key kind, in-order. This use case could make use of an interface like the one exposed here.

I think we'll want to support two separate interfaces for these (and perhaps prohibit range keys from being added through `Add` to avoid accidental misuse). Maybe something like
```
(*Writer) AddInternalRangeKey(key base.InternalKey, value []byte)
```
 and
```
(*Writer) RangeKeySet(start, end, suffix, value []byte)
(*Writer) RangeKeyUnset(start, end, suffix []byte)
(*Writer) RangeKeyDelete(start, end []byte)
```
___
*[sstable/writer.go, line 396 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQXTNDDXT-So96Ndgg:-MqQXTNDDXT-So96Ndgh:b-od32ic) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L396)):*
> ```Go
> 		if !ok {
> 			// We panic here as we should have previously decoded and validated this
> 			// key and value when it was first added to teh range key block.
> ```

nit: s/teh/the/
___
*[sstable/writer.go, line 440 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQXrxv4o7VRjp22Auf:-MqQXrxv4o7VRjp22Aug:behv2te) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L440)):*
> ```Go
> 				// NOTE: the inequality excludes zero, as we allow the end key of the
> 				// lower span be the same as the start key of the upper span, even
> 				// though the range end key is considered an exclusive bound.
> ```

nit: instead of "even though the range end key is considered an exclusive bound," should this say "because the range end key is considered an exclusive bound."? The exclusivity of the end bound is what prevents them from overlapping.
___
*[sstable/writer.go, line 464 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQYFJ34ppiJEsvvcJV:-MqQYFJ34ppiJEsvvcJW:b1aoep1) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L464)):*
> ```Go
> 
> 	// FIXME(travers): More granular properties for RANGEKEYSET, *UNSET, *DEL
> 	// counts?
> ```

I think more granular properties are a good idea: It might be useful for compaction heuristics.

**Review by sumeerbhola — COMMENTED** [review]
> Reviewed 1 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400)* status: 6 of 14 files reviewed, 8 unresolved discussions (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[sstable/writer.go, line 464 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQYFJ34ppiJEsvvcJV:-MqQbQD_DQ8pOoFO_PlZ:b32v57) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L464)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think more granular properties are a good idea: It might be useful for compaction heuristics.
</blockquote></details>

+1

**Review by nicktrav — COMMENTED** [review]
> defer the interface to subsequent PRs.

That would probably be good for directing focus towards internal implementation / external use. I will update this patch to add `(*Writer) AddInternalRangeKey(key base.InternalKey, value []byte)`, as you suggested below. This would be for "internal" use (compactions, etc.).

Then in a follow up, I'll add the per-key-kind functions that cockroach can use, which will also contain the logic to fragment, etc.

> Dismissed @jbowens from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400#-:-MqQhHiX9oZ19rn20ARa:bry2s93)* status: 4 of 14 files reviewed, 2 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[sstable/table.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQFDUa3Il1B79v2gjT:-MqQjjJlE87cGvyHNoJJ:b4c9lyr) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/table.go#L162)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

sgtm
</blockquote></details>

:thumbsup: 
___
*[sstable/writer.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQS8GsCXH22tWjKSNd:-MqQi1l6E6Asd80F8Mr9:bidqepw) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L245)):*
> I think we'll want to support two separate interfaces for these (and perhaps prohibit range keys from being added through Add to avoid accidental misuse)

Sounds good! It should be straight forward to adapt what I have here to make use of the first, which will have more stricter requirements on ordering, etc.

I will tackle the remainder separately, as they will need to make use of the fragmenter, less restrictions on key order, new test cases, etc.
___
*[sstable/writer.go, line 390 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqN6lQZZMyawf8Txp7n-r1-390:-MqQk-HC4VnlOoRdZh9z:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L390)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

Note to self, this comment is old. Update it.
</blockquote></details>

Done.
___
*[sstable/writer.go, line 405 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqN6lQZZMyawf8Txp7n-r1-405:-MqQj9Gw64osHc62p7gO:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L405)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

Note to self, fix the span in the error message.
</blockquote></details>

Done.
___
*[sstable/writer.go, line 440 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQXrxv4o7VRjp22Auf:-MqQj5q34HghUYyH8FKa:b-bw8e14) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L440)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: instead of "even though the range end key is considered an exclusive bound," should this say "because the range end key is considered an exclusive bound."? The exclusivity of the end bound is what prevents them from overlapping.
</blockquote></details>

Updated.
___
*[sstable/writer.go, line 464 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQYFJ34ppiJEsvvcJV:-MqQk5X9C1f925bgvM8w:b-2n3wqj) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L464)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

+1
</blockquote></details>

Great. I'll add counters for the separate key kinds.

**Review by nicktrav — COMMENTED** [review]
This should be good for a proper look now. Have addressed all FIXMEs and exposed the new function on the `Writer` for Pebble-internal range key addition.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400#-:-MqROKI2CwV12-yFTBFy:b1lkjdt)* status: 3 of 14 files reviewed, 2 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[sstable/writer.go, line 464 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqQYFJ34ppiJEsvvcJV:-MqRO7W1A6TJxevu4kZI:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/156a588bcc221ace58075d4dad342a6da54de97d/sstable/writer.go#L464)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

Great. I'll add counters for the separate key kinds.
</blockquote></details>

Done.

**Review by jbowens — COMMENTED** [review]
> Reviewed 1 of 14 files at r1, 1 of 3 files at r2, 2 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400)* status: 7 of 14 files reviewed, 6 unresolved discussions (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[sstable/properties.go, line 120 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVRqhjBkxDDjwptzu4:-MqVRqhjBkxDDjwptzu5:b5wtglf) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/properties.go#L120)):*
> ```Go
> 	NumRangeKeyDels uint64 `prop:"pebble.num.range-key-dels"`
> 	// The number of range keys in this table.
> 	NumRangeKeys uint64 `prop:"pebble.num.range-keys"`
> ```

I think we should drop the aggregate `NumRangeKeys` property, and we can compute it by summing the individual kinds if we need it.
___
*[sstable/writer.go, line 385 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVSTHPDR2812ekr8vr:-MqVSTHPDR2812ekr8vs:bhbvkim) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L385)):*
> ```Go
> 
> // AddInternalRangeKey adds a range key set, unset, or delete key/value pair to
> // the table being writer.
> ```

nit: s/writer/written/
___
*[sstable/writer.go, line 388 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVTJxY0Y5chDwCS1Ki:-MqVTJxY0Y5chDwCS1Kj:bf0gv0j) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L388)):*
> ```Go
> //
> // Range keys must be added ordered by their start key, and in their fragmented
> // span order. Other than spans that are perfectly aligned (same start and end
> ```

Wondering if we should adjust the phrasing to indicate that the `key` parameter, indicating the `Start` boundary, must be strictly ascending.
___
*[sstable/writer.go, line 417 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVUC98A9MIXkHiTyNm:-MqVUC98A9MIXkHiTyNn:bav12pf) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L417)):*
> ```Go
> 		}
> 
> 		switch c := base.InternalCompare(w.compare, prevStartKey, curStartKey); {
> ```

I didn't catch this on my first pass, but this `InternalCompare` comparison should always return `-1` with properly ordered range keys. If the user keys are equal, it will compare on sequence numbers, and if those are equal, it will compare on kind. So the `c == 0` case is always an error case because the user is adding a duplicate key.

We could replace the switch with:
```
if base.InternalCompare(w.compare, prevStartKey, curStartKey) >= 0 {
	w.err = errors.Errorf("pebble: range keys starts must be added in ascending order: %s, %s",
		prevStartKey.Pretty(w.formatKey), key.Pretty(w.formatKey))

}
```
and lift the fragmentation bounds checks from the `default:` case into the main function block.
___
*[sstable/writer.go, line 844 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVY6su4sXBJQoioAQc:-MqVY6su4sXBJQoioAQd:b-zbemhk) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L844)):*
> ```Go
> 		}
> 		w.meta.LargestRangeKey = base.MakeInternalKey(endKey, base.InternalKeySeqNumMax, key.Kind()).Clone()
> 		rangeKeyBH, err = w.writeBlock(w.rangeKeyBlock.finish(), NoCompression)
> ```

@sumeerbhola — do you know why we don't compress the range-del block? I'm wondering if whatever reason also applies here.

Compression is more desirable for range keys, since they carry values but are still fragmented.
___
*[sstable/writer.go, line 1048 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVZIs9344U8fKH_D5H:-MqVZIs9344U8fKH_D5I:b82d28z) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L1048)):*
> ```Go
> 		rangeDelBlock: blockWriter{
> 			restartInterval: 1,
> 		},
> ```

I think we need to initialize the `rangeKeyBlock` field and its `restartInterval` to `1`. This disables key prefix compression. I think we should also update this `blockIter` comment to include range keys in its description of key stability because we'll require that stability for range keys too:

```
// A blockIter provides an additional guarantee around key stability when a
// block has a restart interval of 1 (i.e. when there is no prefix
// compression). Key stability refers to whether the InternalKey.UserKey bytes
// returned by a positioning call will remain stable after a subsequent
// positioning call. The normal case is that a positioning call will invalidate
// any previously returned InternalKey.UserKey. If a block has a restart
// interval of 1 (no prefix compression), blockIter guarantees that
// InternalKey.UserKey will point to the key as stored in the block itself
// which will remain valid until the blockIter is closed. The key stability
// guarantee is used by the range tombstone code which knows that range
// tombstones are always encoded with a restart interval of 1. This per-block
// key stability guarantee is sufficient for range tombstones as they are
// always encoded in a single block.
```

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400)* status: 7 of 14 files reviewed, 6 unresolved discussions (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[sstable/writer.go, line 844 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVY6su4sXBJQoioAQc:-MqVbGB564QQygZdye6w:b-mt113b) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L844)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

@sumeerbhola — do you know why we don't compress the range-del block? I'm wondering if whatever reason also applies here.

Compression is more desirable for range keys, since they carry values but are still fragmented.
</blockquote></details>

Hmm, I suspect it's because we cache compressed blocks, so we want to avoid the CPU cost of decompressing the range-del block on every read in the sstable. That concern also applies to range keys, because every MVCC read will need to consult the range-key block if there is one...

I think we should keep it uncompressed for now, but I think this deserves more thought.

**Review by jbowens — COMMENTED** [review]
Just thought of something—We'll want to feed these range keys to the block property collectors, but probably through a new range-key specific method on the `BlockPropertyCollector`. Some block-level property filters might want to adjust the table-level properties based on the range keys.

I think tackling that in a followup PR is fine too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400#-:-MqVqsEFBM0FCa3A80i2:bajsknf)* status: 7 of 14 files reviewed, 6 unresolved discussions (waiting on @jbowens, @nicktrav, and @sumeerbhola)

**Review by nicktrav — COMMENTED** [review]
I added an item to the list for wiring up the block property collectors.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400#-:-MqolxuwB521h1bFiaOv:b39771d)* status: 7 of 14 files reviewed, 5 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[sstable/properties.go, line 120 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVRqhjBkxDDjwptzu4:-Mqom_yv-_hCmih1tk1m:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/properties.go#L120)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we should drop the aggregate `NumRangeKeys` property, and we can compute it by summing the individual kinds if we need it.
</blockquote></details>

Done.
___
*[sstable/writer.go, line 388 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVTJxY0Y5chDwCS1Ki:-MqouPUW4OhbQB9l6jPP:b8f0rkp) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L388)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Wondering if we should adjust the phrasing to indicate that the `key` parameter, indicating the `Start` boundary, must be strictly ascending.
</blockquote></details>

Done. Matched the verbiage that we use elsewhere.
___
*[sstable/writer.go, line 417 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVUC98A9MIXkHiTyNm:-Mqp3zs4Ei7hapOEGALA:bgmi00r) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L417)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I didn't catch this on my first pass, but this `InternalCompare` comparison should always return `-1` with properly ordered range keys. If the user keys are equal, it will compare on sequence numbers, and if those are equal, it will compare on kind. So the `c == 0` case is always an error case because the user is adding a duplicate key.

We could replace the switch with:
```
if base.InternalCompare(w.compare, prevStartKey, curStartKey) >= 0 {
	w.err = errors.Errorf("pebble: range keys starts must be added in ascending order: %s, %s",
		prevStartKey.Pretty(w.formatKey), key.Pretty(w.formatKey))

}
```
and lift the fragmentation bounds checks from the `default:` case into the main function block.
</blockquote></details>

Ah nice catch. That was left over from before making use of `base.InternalCompare`. Cleaned this up.
___
*[sstable/writer.go, line 844 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVY6su4sXBJQoioAQc:-Mqp5K2T2RNEd3NTW749:b-7yd1q4) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L844)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Hmm, I suspect it's because we cache compressed blocks, so we want to avoid the CPU cost of decompressing the range-del block on every read in the sstable. That concern also applies to range keys, because every MVCC read will need to consult the range-key block if there is one...

I think we should keep it uncompressed for now, but I think this deserves more thought.
</blockquote></details>

Added a TODO.
___
*[sstable/writer.go, line 1048 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1400#-MqVZIs9344U8fKH_D5H:-Mqp7AkKA54P9qJWH4GQ:b-gd02e6) ([raw file](https://github.com/cockroachdb/pebble/blob/428046ba6540e9e932880cfc7e1df1b08de4576c/sstable/writer.go#L1048)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we need to initialize the `rangeKeyBlock` field and its `restartInterval` to `1`. This disables key prefix compression. I think we should also update this `blockIter` comment to include range keys in its description of key stability because we'll require that stability for range keys too:

```
// A blockIter provides an additional guarantee around key stability when a
// block has a restart interval of 1 (i.e. when there is no prefix
// compression). Key stability refers to whether the InternalKey.UserKey bytes
// returned by a positioning call will remain stable after a subsequent
// positioning call. The normal case is that a positioning call will invalidate
// any previously returned InternalKey.UserKey. If a block has a restart
// interval of 1 (no prefix compression), blockIter guarantees that
// InternalKey.UserKey will point to the key as stored in the block itself
// which will remain valid until the blockIter is closed. The key stability
// guarantee is used by the range tombstone code which knows that range
// tombstones are always encoded with a restart interval of 1. This per-block
// key stability guarantee is sufficient for range tombstones as they are
// always encoded in a single block.
```
</blockquote></details>

Good catch. Updated.

**Review by jbowens — APPROVED** [review]
:lgtm:!

> Reviewed 1 of 14 files at r1, 3 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400#-:-MqpDiZB-abtTHQ9Aqs0:bg7igkw)* status: 8 of 15 files reviewed, all discussions resolved (waiting on @jbowens and @sumeerbhola)

**Comment by nicktrav** [comment]
@sumeerbhola and I chatted offline - I'll merge this as-is, and will follow-up on any comments that come up post-merge.

TFTRs!

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400)* status: 7 of 15 files reviewed, 1 unresolved discussion

___
*[sstable/table.go, line 89 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1400#-Mqzf3YMAsUNQJgFoG-S:-Mqzf3YMAsUNQJgFoG-T:b-38lmqd) ([raw file](https://github.com/cockroachdb/pebble/blob/fde3810aed483ac3a7bf7bf6c52c249e59d3aec1/sstable/table.go#L89)):*
> ```Go
> [index block] (for single level index)
> [meta range key block] (optional)
> [meta properties block]
> ```

The meta properties block is after the rangedel block

**Review by nicktrav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1400)* status: 7 of 15 files reviewed, 1 unresolved discussion

___
*[sstable/table.go, line 89 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1400#-Mqzf3YMAsUNQJgFoG-S:-MqzkPcz0jQ1ENAPKBdC:b-kezlea) ([raw file](https://github.com/cockroachdb/pebble/blob/fde3810aed483ac3a7bf7bf6c52c249e59d3aec1/sstable/table.go#L89)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The meta properties block is after the rangedel block
</blockquote></details>

Woops! Thanks for catching. Fix: #1413.

**Pull request merged**